internal/cluster/certificates.go

```diff
@@ -0,0 +1,371 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cluster
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/hex"
+	"fmt"
+	"math/big"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/cert"
+	bootstrapv1 "sigs.k8s.io/cluster-api-bootstrap-provider-kubeadm/api/v1alpha2"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha2"
+	"sigs.k8s.io/cluster-api/util/certs"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	rootOwnerValue = "root:root"
+
+	// TLSKeyDataName is the name of the secret key within the certificate secret of the key
+	TLSKeyDataName = "tls.key"
+
+	// TLSCrtDataName is the name of the secret key within the certificate secret of the certificate
+	TLSCrtDataName = "tls.crt"
+
+	// ClusterCAName is the secret name suffix for apiserver CA
+	ClusterCAName = "ca"
+
+	// EtcdCAName is the secret name suffix for the Etcd CA
+	EtcdCAName = "etcd"
+
+	// ServiceAccountName is the secret name suffix for the Service Account keys
+	ServiceAccountName = "sa"
+
+	// FrontProxyCAName is the secret name suffix for Front Proxy CA
+	FrontProxyCAName = "proxy"
+)
+
+// Certificates are the certificates necessary to bootstrap a cluster.
+type Certificates []*Certificate
+
+// NewCertificates return an initialized but empty set of CA certificates needed to bootstrap a cluster.
+func NewCertificates() Certificates {
+	return Certificates{
+		&Certificate{Name: ClusterCAName},
+		&Certificate{Name: EtcdCAName},
+		&Certificate{Name: ServiceAccountName},
+		&Certificate{Name: FrontProxyCAName},
+	}
+}
+
+// GetCertificateByName returns a certificate by the given name.
+// This could be removed if we use a map instead of a slice to hold certificates, however other code becomes more complex.
+func (c Certificates) GetCertificateByName(name string) *Certificate {
+	for _, certificate := range c {
+		if certificate.Name == name {
+			return certificate
+		}
+	}
+	return nil
+}
+
+// GetCertificates looks up each certificate from secrets and populates the certificate with the secret data.
+func (c Certificates) GetCertificates(ctx context.Context, ctrlclient client.Client, cluster *clusterv1.Cluster) error {
+	// Look up each certificate as a secret and populate the certificate/key
+	for _, certificate := range c {
+		secret := &corev1.Secret{}
+		key := client.ObjectKey{
+			Name:      certificate.SecretName(cluster.Name),
+			Namespace: cluster.Namespace,
+		}
+		if err := ctrlclient.Get(ctx, key, secret); err != nil {
+			if apierrors.IsNotFound(err) {
+				kp, err := generateCACert()
```

Comment:
Do you want this function (GetCertificates) to generate if the secret isn't there? We have GetOrCreateCertificates that handles that case?

Reply:
oof yes, this was my first strategy but I broke it out as you saw. 

I must have not removed this bit, thanks for the catch. I'll update and retest.

Reply:
pushing this change then adding tests to the certificates package